presto-main/src/main/java/com/facebook/presto/sql/relational/SqlToRowExpressionTranslator.java

```diff
@@ -528,64 +528,77 @@ protected RowExpression visitCoalesceExpression(CoalesceExpression node, Void co
         @Override
         protected RowExpression visitSimpleCaseExpression(SimpleCaseExpression node, Void context)
         {
-            ImmutableList.Builder<RowExpression> arguments = ImmutableList.builder();
-
-            arguments.add(process(node.getOperand(), context));
-
-            for (WhenClause clause : node.getWhenClauses()) {
-                arguments.add(specialForm(
-                        WHEN,
-                        getType(clause),
-                        process(clause.getOperand(), context),
-                        process(clause.getResult(), context)));
-            }
-
-            Type returnType = getType(node);
-
-            arguments.add(node.getDefaultValue()
-                    .map((value) -> process(value, context))
-                    .orElse(constantNull(returnType)));
-
-            return specialForm(SWITCH, returnType, arguments.build());
+            return buildSwitch(process(node.getOperand(), context), node.getWhenClauses(), node.getDefaultValue(), getType(node), context);
         }
 
         @Override
         protected RowExpression visitSearchedCaseExpression(SearchedCaseExpression node, Void context)
         {
-            /*
-                Translates an expression like:
+            // Rewriter situations where it's CASE WHEN x = v1 THEN r1, WHEN x = v2 THEN r2 ... END to:
+            // CASE x WHEN v1 THEN r1 WHEN v2 THEN r2 ...
+            // So that better code is generated espeically in tableau
+            RowExpression lhsArg = null;
+            for (WhenClause whenClause : node.getWhenClauses()) {
+                Expression predicate = whenClause.getOperand();
+                if (ExpressionTreeUtils.isEqualComparisonExpression(predicate) ||
+                        ExpressionTreeUtils.isInValuesComparisonExpression(predicate)) {
+                    RowExpression operand = process(predicate.getChildren().get(0), context);
+                    if (lhsArg != null && !operand.equals(lhsArg)) {
+                        lhsArg = null;
+                        break;
+                    }
+                    lhsArg = operand;
+                }
+                else {
+                    lhsArg = null;
+                    break;
+                }
+            }
 
-                    case when cond1 then value1
-                         when cond2 then value2
-                         when cond3 then value3
-                         else value4
-                    end
+            if (lhsArg != null) {
+                // We found the pattern. So rebuild the When list to just include the rhs of the equals.
+                ImmutableList.Builder<WhenClause> newWhenClauses = ImmutableList.builder();
+                for (WhenClause whenClause : node.getWhenClauses()) {
+                    if (ExpressionTreeUtils.isInValuesComparisonExpression(whenClause.getOperand())) {
```

Comment:
static import

Reply:
done. I think you mean isInValuesComparisonExpression